src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/SelfDiagnostics/ApplicationInsightsEventListener.cs

```diff
@@ -0,0 +1,123 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Timers;
+
+namespace Microsoft.Azure.WebJobs.Logging.ApplicationInsights
+{
+    /// <summary>    
+    /// Implementation of <see cref="EventListener"/> that listens to events produced by ApplicationInsights SDK.
+    /// Logs data either every 10 seconds or when the batch becomes full, whichever occurs first.
+    /// Logs in batch to reduce the volume of logs in kusto    
+    /// </summary>
+    internal class ApplicationInsightsEventListener : EventListener
+    {
+        private static readonly DiagnosticListener _source = new DiagnosticListener(string.Concat(ApplicationInsightsDiagnosticConstants.DiagnosticSourcePrefix, nameof(ApplicationInsightsEventListener)));
+        private readonly EventLevel _eventLevel;
+
+        private const int LogFlushIntervalMs = 10 * 1000;
+        private const string EventSourceNamePrefix = "Microsoft-ApplicationInsights-";
+        private const string EventName = nameof(ApplicationInsightsEventListener);
+        private const int MaxLogLinesPerFlushInterval = 30;
+
+        private Timer _flushTimer;
+        private ConcurrentQueue<string> _logBuffer = new ConcurrentQueue<string>();
+        private ConcurrentQueue<EventSource> _eventSource = new ConcurrentQueue<EventSource>();
+        private static object _syncLock = new object();
+        private bool _disposed = false;
+
+        public ApplicationInsightsEventListener(EventLevel eventLevel)
+        {
+            this._eventLevel = eventLevel;
+            _flushTimer = new Timer
+            {
+                AutoReset = true,
+                Interval = LogFlushIntervalMs
+            };
+            _flushTimer.Elapsed += (sender, e) => Flush();
+            _flushTimer.Start();
+        }
+
+        protected override void OnEventSourceCreated(EventSource eventSource)
+        {
+            if (eventSource.Name.StartsWith(EventSourceNamePrefix))
+            {
+                EnableEvents(eventSource, _eventLevel, EventKeywords.All);
+                _eventSource.Enqueue(eventSource);
+            }
+            base.OnEventSourceCreated(eventSource);
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (!string.IsNullOrWhiteSpace(eventData.Message))
+            {
+                _logBuffer.Enqueue(string.Format(CultureInfo.InvariantCulture, eventData.Message, eventData.Payload.ToArray()));
+                if (_logBuffer.Count >= MaxLogLinesPerFlushInterval)
+                {
+                    Flush();
+                }
+            }
+        }
+
+        public void Flush()
+        {
+            if (_logBuffer.Count == 0)
+            {
+                return;
+            }
+            lock (_syncLock)
+            {
```

Comment:
Check for `_logBuffer.Count == 0` again once you have the lock? Another thread could have potentially flushed everything already...